### ISSUE TITLE
Allow more characters in snapshot hames

### ISSUE THREAD

**Issue by Phantop** [elfshaker/elfshaker#62]
Particularly, I think that allowing . and spaces for snapshot names would be useful because many programs use the former to denote release versions and the latter could allow more descriptive snapshot names, if so desired.

**Comment by veselink1** [comment]
I'd be happy if we could allow more characters in snapshot tags. I see the value in adding '.' in particular. I've concerns for allowing spaces through - handling spaces "correctly" can be difficult: 
- It is unclear to me how the output of `elfshaker list` should be formatted when there are spaces present. We would like it to be easy to parse from shell scripts, and we currently separate the pack and snapshot names by spaces. 
- We have a couple of bash scripts https://github.com/elfshaker/elfshaker/tree/main/contrib which I believe should handle an extra space in snapshot tags correctly, but I'm not 100% sure. I am concerned that we might end up breaking them, so additional validation will be needed before we allow that. 

The full list of allowed characters currently is: https://github.com/elfshaker/elfshaker/blob/cb127c1641ab0866f263fa029155c8911c28a523/src/repo/pack.rs#L32

**Comment by peterwaller-arm** [comment]
Yes, let's add `.`! Contributions welcome, or we might get around to this when we have time (but it might be months++).

I'm quite nervous about allowing space, for the aforementioned reason: it makes machine-handling of outputs more error-prone, so I currently do not favour this unless there is a compelling reason + story around handling them gracefully.

While we're considering characters I can see the importance in allowing a much wider range of characters, to enable this to be useful in other languages for example. I'm not immediately clear how to do this in a safe way which avoids introducing subtle bugs or usability hazards, so help is wanted analysing this before we go too wide.

In the mean time, adding specific characters which are clearly safe is a no-brainer, please feel free to submit those as pull requests as-needed.

**Comment by Laura7089** [comment]
Are there any complications to this beyond just adding `.` to the `EXTRA_ID_CHARS` array?

**Comment by veselink1** [comment]
We use `EXTRA_ID_CHARS` for both packs and snapshots and packs have 1-1 correspondence with files on the filesystem (minus the file extension). 

If `.` is also desirable for pack names (probably not?), it can become a tiny amount more involved. With `.` in particular, the file extension handling code might need to be reviewed and potentially fixed. I know for a fact that we are using `set_extension` on the bare filenames as an append operation. I imagine this will create problems, but can be easily recast as a string concat. 

If we don't want to change the rules for pack names, `EXTRA_ID_CHARS` might have to be split into two arrays, one for packs and one for snapshots.

**Comment by veselink1** [comment]
I agree it is better to keep the pack and snapshot naming rules the same. Allowing `.` in pack names just requires a bit more work, but is definitely doable. I'm merely touching on some of the issues with going either way, in case you or anyone else wants to work on this issue.

**Comment by Mistuke** [comment]
This would be helpful for me as well, my pack names contain the output of `git describe` which when when close to a release tag will have a version number.

**Comment by peterwaller-arm** [comment]
I would like to address this for 1.0.0 which I hope to release in the near future.

**Comment by peterwaller-arm** [comment]
Better late than never, I'm planning to make the spec the below for 1.0. If you think you might have feedback please pipe up or email me / message me, I intend to implement this soon but will delay if there is feedback. I plan to lean on the filesystem enforcing these rules, which means that loose directories won't necessarily be portable across systems/filesystems by copying and pasting, but I think that's fine as the intended mechanism of sharing is to use pack files, and snapshot names don't otherwise appear on the filesystem.

## Rules

A tag is one or more /-separated components.

- Allow Unicode letters, marks, numbers, punctuation, symbols, and internal ASCII spaces.
- Allow . within components, including leading dots.
- Reject empty components, . and ...
- Reject leading/trailing spaces and trailing periods.
- Reject :, \, <, >, ", |, ?, *, controls, invisible Unicode categories, and non-ASCII whitespace.
- Reject Windows device names case-insensitively, including CON, PRN, AUX, NUL, COM1–COM9, and LPT1–LPT9, including names with extensions.
- Reject non-final components ending in .pack.idx to avoid file/directory collisions.
- Preserve exact case and Unicode spelling; perform no normalisation.

Examples:

- Valid: v1.2.3, release 1.2, llvm/18.1.2, .candidate, résumé/β.2.
- Invalid: ., .., a/../b, a//b, a:b, a\b, a., foo.pack.idx/bar, names containing tabs or newlines.

## Consequences and Implementation

- Store raw loose indexes as packs/loose/<tag>.pack.idx, creating hierarchy directories as needed.
- Append .pack.idx literally; replace all extension-replacement operations that truncate dotted names.
- Keep existing [A-Za-z0-9_-]+ paths unchanged and retain the current pack-index format.
- Separate regular pack-name validation from snapshot-tag validation. Regular pack names retain their current charset; internal loose/<tag> IDs use the snapshot rules.

- Harden every pack path against absolute paths, empty components, and traversal—even when callers construct PackId directly.
- Validate tags consistently in SnapshotId::new, parsing, HEAD, and repository operations.
- Before overwriting a loose index, verify that its stored tag exactly matches. A case-folding or normalisation collision becomes an error, never silent data loss.
- Reconstruct loose logical IDs from the tag stored inside their index where filesystem spelling may differ.
- Remove empty hierarchy directories after loose-index deletion.

Raw storage deliberately inherits filesystem behaviour:

- Case or Unicode-equivalent tags may coexist on Linux but collide on Windows or some macOS filesystems.
- Loose data directories relying on those distinctions are not reliably portable. Packing snapshots before transfer avoids filename-level tag collisions.
- Filesystem component and total-path length limits apply.
- Shell punctuation requires quoting.
- Leading-dot components may produce hidden paths on Unix.
